src/DynamoPackages/PackageManagerExtension.cs

```diff
@@ -149,6 +176,87 @@ private void OnMessageLogged(ILogMessage msg)
                 this.MessageLogged(msg);
             }
         }
+        
+        private void OnCurrentWorkspaceChanged(IWorkspaceModel ws)
+        {
+            if (ws is WorkspaceModel)
+            {
+                if (currentWorkspace != null)
+                {
+                    (currentWorkspace as WorkspaceModel).CollectingCustomNodePackageDependencies -= GetCustomNodePackageFromID;
+                    (currentWorkspace as WorkspaceModel).CollectingNodePackageDependencies -= GetNodePackageFromAssemblyName;
+                }
+                
+                (ws as WorkspaceModel).CollectingCustomNodePackageDependencies += GetCustomNodePackageFromID;
+                (ws as WorkspaceModel).CollectingNodePackageDependencies += GetNodePackageFromAssemblyName;
+                currentWorkspace = ws;
+            }
+        }
+
+        private PackageInfo GetNodePackageFromAssemblyName(AssemblyName assemblyName)
+        {
+            if (NodePackageDictionary.ContainsKey(assemblyName.FullName))
+            {
+                return NodePackageDictionary[assemblyName.FullName];
+            }
+            return null;
+        }
+
+        private PackageInfo GetCustomNodePackageFromID(Guid functionID)
+        {
+            if (CustomNodePackageDictionary.ContainsKey(functionID))
+            {
+                return CustomNodePackageDictionary[functionID];
+            }
+            return null;
+        }
+
+        private void OnPackageLoaded(Package package)
+        {
+            // Create NodePackageDictionary if it doesn't exist
+            if (NodePackageDictionary == null)
+            {
+                NodePackageDictionary = new Dictionary<string, PackageInfo>();
+            }
+            // Add new assemblies to NodePackageDictionary
+            foreach (var assembly in package.LoadedAssemblies.Select(a => AssemblyName.GetAssemblyName(a.Assembly.Location)))
+            {
+                NodePackageDictionary[assembly.FullName] = new PackageInfo(package.Name, package.VersionName);
+            }
+
+            // Create CustomNodePackageDictionary if it doesn't exist
+            if (CustomNodePackageDictionary == null)
+            {
+                CustomNodePackageDictionary = new Dictionary<Guid, PackageInfo>();
+            }
+            // Add new custom nodes to CustomNodePackageDictionary
+            foreach (var cn in package.LoadedCustomNodes)
+            {
+                var pInfo = new PackageInfo(package.Name, package.VersionName);
+                CustomNodePackageDictionary[cn.FunctionId] = pInfo;
```

Comment:
same issue here - what if two packages expose the same custom node... I'm not sure this is possible - BUT it's definitely possible for one package to republish another package's custom nodes - I am not sure if `LoadedCustomNodes` will have those dependent nodes included though.